ssl/statem/statem_srvr.c

```diff
@@ -3374,28 +3374,26 @@ static int tls_process_cke_gost(SSL *s, PACKET *pkt)
                  SSL_R_DECRYPTION_FAILED);
         goto err;
     }
-    if (asn1len == 0x81) {
+    if (asn1len >= 0x81) {
         /*
-         * Long form length. Should only be one byte of length. Anything else
-         * isn't supported.
-         * We did a successful peek before so this shouldn't fail
+         * Long form length.
          */
-        if (!PACKET_forward(pkt, 1)) {
+        size_t length = asn1len & 0x7F;
+        if (!PACKET_forward(pkt, length)) {
```

Comment:
This doesn't look right. Previously we supported two formats:

1) Long form with length with a single byte of length. In this case `pkt` looks like this:

`0x81, len (1 byte), len bytes of data`


We then moved forward by 1 byte so that pkt then looks like:

`len (1 byte), len bytes of data`

2) Short form length

In that case pkt looked like

`len (1 byte), len bytes of data`

In both cases `pkt` ended up with 1 byte of length data followed by the actual data, and therefore we could use the `PACKET_as_length_prefixed_1` call to read it. That expects exactly 1 byte of length data, followed by the actual data (which should extend to the end of the packet).

With these new changes lets say we have a packet with long form length where `asn1len == 0x82`. In this case `pkt` will look like this:

`0x82, hi-byte-of-len, lo-byte-of-len, len bytes of data`

In this example `length` will have the value 2 to indicate there are 2 bytes of length data. So after the forward we will have:

`lo-byte-of-len, len bytes of data`

This means that calling `PACKET_as_length_prefixed_1` will fail because we will only read part of the data. Since that call expects us to consume *all* the data in the packet it fails. Changing it to PACKET_get_length_prefixed_1 doesn't really solve the problem. It just stops the call from failing. It still doesn't read all the data, and doesn't consume all the PACKET - it just throws it away instead.

Reply:
Yes, it's the desired behavior. Here is the KX blob struct:
```
  0 509: SEQUENCE {
  4 239:   SEQUENCE {
  7  40:     SEQUENCE {
  9  32:       OCTET STRING
       :         47 D8 50 C0 57 9D 30 1A 51 50 F8 7D 0B AF C5 EC
       :         C3 DE 46 85 A4 95 A2 6E 01 5A A3 1A 85 6D 0F F4
 43   4:       OCTET STRING 82 8B 09 71
       :       }
 49 194:     [0] {
 52   9:       OBJECT IDENTIFIER '1 2 643 7 1 2 5 1 1'
 63 170:       [0] {
 66  33:         SEQUENCE {
 68   8:           OBJECT IDENTIFIER '1 2 643 7 1 1 1 2'
 78  21:           SEQUENCE {
 80   9:             OBJECT IDENTIFIER '1 2 643 7 1 2 1 2 1'
 91   8:             OBJECT IDENTIFIER '1 2 643 7 1 1 2 3'
       :             }
       :           }
101 132:         BIT STRING, encapsulates {
105 128:           OCTET STRING
       :             8B 69 FC EC 67 D2 46 C6 A5 1F 25 D2 37 19 3E 27
       :             B5 F9 7E E5 37 E7 9B 65 76 11 AE DD EB E3 21 21
       :             B8 69 A0 EC 0B F4 A9 B6 5F 5F 26 14 07 5D 1F DA
       :             36 43 F0 76 B4 52 56 64 DF FF C4 6C 23 0E 7B 1A
       :             D8 84 20 59 69 34 4B 2D C2 E9 0C 2A CA 68 E7 EA
       :             0D 2D 88 93 9E 93 2C 0F F2 E4 2C F1 B5 F2 78 76
       :             C0 83 DF 9B D4 37 55 4C C4 1C 3A 82 0D B8 CC 02
       :             1A B1 51 C3 DF A1 4A 0D 52 36 7E BD 6A 86 D9 C4
       :           }
       :         }
236   8:       OCTET STRING 8F AA 92 02 D8 F2 C0 D3
       :       }
       :     }
246 263:   SEQUENCE {
250 259:     SEQUENCE {
254  65:       SEQUENCE {
256  40:         SEQUENCE {
258  32:           OCTET STRING
       :             47 D8 50 C0 57 9D 30 1A 51 50 F8 7D 0B AF C5 EC
       :             C3 DE 46 85 A4 95 A2 6E 01 5A A3 1A 85 6D 0F F4
292   4:           OCTET STRING 82 8B 09 71
       :           }
298  21:         [0] {
300   9:           OBJECT IDENTIFIER '1 2 643 7 1 2 5 1 1'
311   8:           OCTET STRING 8F AA 92 02 D8 F2 C0 D3
       :           }
       :         }
321 189:       OCTET STRING, encapsulates {
324 186:         OCTET STRING
       :           ':Licinfo_itcs:dcgGN+4vmMNL/Yc1pRtKSWmvUv6xu5CZKT'
       :           'JTMYvPktB0ui0KN3BWkuR/nkUhwCiVifkJhEIg3wpHt1rEI+'
       :           'JQNkTEr1oQN+vxZmDQdOPJNC0CABwWQg2lcJloPayYJUM3Z9'
       :           'pWDVrc18o0+Wg4sOzZarEYprOJWePs7skFGH5/1AA='
       :         }
       :       }
       :     }
       :   }
```
So we take the 1st element of the sequence containing the wrapped key, and ignore the rest as it does not contain any information necessary for us.

Reply:
What?? And that *works*? So we end up using what ever random length is in the lo-byte (which presumably could be anything from 0 to 255) and that's the amount of data we feed to the following EVP_PKEY_decrypt()? Surely that will just fail unless you happen to pass it the correct length?

Reply:
It works, but I will rewrite it using a proper ASN1 parser.
Turned to WIP.

Reply:
One option is to add a new packet function...something like `PACKET_as_der_length_prefixed()` or similar.

Reply:
I always disliked this piece of code, both for a client and a server, but never had a reason to rewrite it.

https://tools.ietf.org/html/draft-chudov-cryptopro-cptls-04#section-3.6 describes a basic form of the message. The OPTIONAL element was never supported in practice, but formally leaves a chance for implementors to add extra information. So I'll try to extract the necessary part using i2d code, and the GOST engine invokes i2d function for the extracted blob, as it does now.